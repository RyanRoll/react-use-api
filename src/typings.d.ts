/// <reference types="react" />

import * as Axios from 'axios'

export = ReactUseApi
export as namespace ReactUseApi

declare namespace ReactUseApi {
  type Settings = typeof import('./common').defaultSettings
  type CustomSettings = Partial<Settings>
  type ACTIONS = typeof import('./common').ACTIONS
  type InitState = typeof import('./common').initState
  type SingleConfig = Axios.AxiosRequestConfig
  type MultiConfigs = SingleConfig[]
  type Config = SingleConfig | MultiConfigs
  type ApiResponse = Axios.AxiosResponse<JsonObject>
  type SingleData = JsonObject | undefined | any
  type Data = SingleData | SingleData[] | undefined | any
  type RequestFn = (
    cfg?: ReactUseApi.Config,
    keepState?: boolean
  ) => Promise<any>

  interface Context {
    settings?: Settings
    isSSR?: boolean
    collection?: SSRCollection
    isConfigured?: boolean
    renderSSR?: Settings['renderSSR']
    clearCache?: () => void
  }
  interface CustomContext extends Omit<Context, 'settings'> {
    settings?: CustomSettings
  }
  interface SSRCollection {
    ssrConfigs: SSRConfigs[]
    cacheKeys: Set<string>
  }
  interface SSRConfigs {
    config: Config
    cacheKey: string
  }
  interface ApiProviderProps {
    context?: CustomContext
  }
  interface Options {
    watch?: any[]
    dependencies?: dependencies
    skip?: boolean
    useCache?: boolean
    $cacheKey?: string
    $hasChangedConfig?: boolean
    handleData?: (data: Data, newState: State) => any
    shouldRequest?: () => boolean | void
  }
  interface dependencies {
    readonly [key: string]: any
  }

  interface CacheData {
    response?: ApiResponse | ApiResponse[]
    error?: Axios.AxiosError
  }
  interface Action extends CacheData {
    type: string
    options?: Options
    fromCache?: boolean
  }
  interface State extends InitState, CacheData, JsonObject {
    data?: Data
    prevData?: Data
    prevState?: State
    dependencies?: Options['dependencies']
  }

  interface RefData {
    id: number
    cacheKey: string
    state: State
    config: Config
    refreshFlag: number
    isInit: boolean
    isRequesting: boolean
    hasFed: boolean
    timeoutID: number
    useCache: boolean
    // options: Options // debug only
  }

  interface DispatchClusterElement {
    id: number
    dispatch: React.Dispatch<Action>
    refData: RefData
  }

  interface RECORDS {
    [cacheKey: string]: {
      dispatches: DispatchClusterElement[]
      suspense?: {
        promise: Promise<DataResponse>
        done: boolean
        reference: number
      }
    } & CacheData
  }

  type MiddlewareInterrupt = () => void
  interface Middleware {
    onHandleOptions?: (
      config?: Config,
      args?: { options: Options },
      interrupt?: MiddlewareInterrupt
    ) => Options
    onStart?: (
      config?: Config,
      args?: { state: State },
      interrupt?: MiddlewareInterrupt
    ) => State
    onHandleData?: (
      config?: Config,
      args?: {
        data: Data
        response: DataResponse
      },
      interrupt?: MiddlewareInterrupt
    ) => Data
    onHandleError?: (
      config?: Config,
      args?: { error: ErrorResponse },
      interrupt?: MiddlewareInterrupt
    ) => any
    onSuccess?: (
      config?: Config,
      args?: {
        data: Data
        response: DataResponse
      },
      interrupt?: MiddlewareInterrupt
    ) => void
    onError?: (
      config?: Config,
      args?: {
        error: ErrorResponse
        response: DataResponse
      },
      interrupt?: MiddlewareInterrupt
    ) => void
    onDone: (
      config?: Config,
      args?: {
        data: Data
        error: ErrorResponse
        response: DataResponse
      },
      interrupt?: MiddlewareInterrupt
    ) => void
  }
  type Middlewares = Middleware[]

  interface JsonObject {
    [key: string]: any
  }
}
