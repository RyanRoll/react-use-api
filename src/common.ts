import axios, { AxiosStatic, AxiosInstance, AxiosError } from 'axios'
import LRU from 'lru-cache'

// export const cacheKeySymbol: unique symbol = Symbol('cacheKey') TS still not friendly to symbol...
export const defaultSettings = {
  cache: new LRU<string, ReactUseApi.CacheData | any>(),
  axios: axios as AxiosStatic | AxiosInstance,
  maxRequests: 100, // max requests count when running ReactDom.renderToString in SSR
  useCacheData: true, // whether to use the cached api data come from server
  alwaysUseCache: false, // whether to use the cached api data always for each api call
  clearLastCacheWhenConfigChanges: true, // clear the last cache data with the last config when the config changes
  debug: false,
  clientCacheVar: '__USE_API_CACHE__', // the property name of window to save the cached api data come from server side
  isSSR: (...args: any[]): boolean | void => typeof window === 'undefined',
  renderSSR: (...args: any[]): string => '', // a callback to render SSR HTML string
  shouldUseApiCache: (
    config?: ReactUseApi.Config,
    cacheKey?: string
  ): boolean | void => true, // a callback to decide whether to use the cached api data
}
export const ACTIONS = {
  REQUEST_START: 'REQUEST_START',
  REQUEST_END: 'REQUEST_END',
}
export const initState = {
  loading: false,
  fromCache: false,
  $cacheKey: '',
}

export const configure = (
  context: ReactUseApi.CustomContext,
  isSSR = false
) => {
  if (context.isConfigured) {
    return context as ReactUseApi.Context
  }
  const { settings: custom } = context
  const settings = { ...defaultSettings }
  if (isObject(custom)) {
    Object.keys(custom).forEach((key: keyof ReactUseApi.Settings) => {
      const value = custom[key]
      if (defaultSettings.hasOwnProperty(key) && !isNil(value)) {
        // @ts-ignore
        settings[key] = value
      }
    })
  }
  isSSR =
    isSSR !== true && isFunction(settings.isSSR) ? !!settings.isSSR() : isSSR
  Object.assign(context, {
    settings,
    isSSR,
    isConfigured: true,
    collection: {
      ssrConfigs: [],
      cacheKeys: new Set<string>(),
    } as ReactUseApi.SSRCollection,
    clearCache() {
      settings?.cache?.reset()
    },
  })
  return context as ReactUseApi.Context
}

export async function axiosAll(
  context: ReactUseApi.Context,
  config: ReactUseApi.Config
): Promise<ReactUseApi.ApiResponse>
export async function axiosAll(
  context: ReactUseApi.Context,
  config: ReactUseApi.MultiConfigs
): Promise<ReactUseApi.ApiResponse[]>
export async function axiosAll(
  context: ReactUseApi.Context,
  config: ReactUseApi.Config | ReactUseApi.MultiConfigs
): Promise<ReactUseApi.ApiResponse | ReactUseApi.ApiResponse[]> {
  const {
    settings: { axios: client },
  } = context
  const isMulti = Array.isArray(config)
  const requests = ((isMulti
    ? config
    : [config]) as ReactUseApi.MultiConfigs).map((cfg) => client(cfg))
  try {
    const responses = await Promise.all<ReactUseApi.ApiResponse>(requests)
    responses.forEach(tidyResponse)
    const response = responses.length === 1 ? responses[0] : responses
    return response
  } catch (error) {
    const { response } = error as ReactUseApi.CacheData['error']
    if (response) {
      tidyResponse(response)
    }
    throw error as AxiosError
  }
}

// for cache
export const tidyResponse = (response: ReactUseApi.ApiResponse) => {
  if (response) {
    delete response.config
    delete response.request
  }
  return response
}

export const getResponseData = (
  options: ReactUseApi.Options,
  state: ReactUseApi.State
) => {
  const { response } = state
  const isMultiApis = Array.isArray(response)
  let data = isMultiApis
    ? (response as ReactUseApi.ApiResponse[]).map((each) => each.data)
    : (response as ReactUseApi.ApiResponse).data
  const { handleData } = options
  if (isFunction(handleData)) {
    data = handleData(data, state)
  }
  return data
}

export const isObject = (target: any) =>
  !!target && Object.prototype.toString.call(target) === '[object Object]'

export const isFunction = (target: any) =>
  !!target && typeof target === 'function'

export const isNil = (value: any) =>
  value === undefined || value === null || value === ''
