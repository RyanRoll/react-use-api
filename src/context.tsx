import { createContext } from 'react'

export const ApiContext = createContext<ReactUseApi.Context>({})
