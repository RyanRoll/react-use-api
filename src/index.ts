export { default, useApi } from './useApi'
export { injectSSRHtml, loadApiCache } from './ssr'
export { ApiProvider } from './ApiProvider'
export { ApiContext } from './context'
