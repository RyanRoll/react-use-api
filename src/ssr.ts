import { configure, axiosAll, defaultSettings, isFunction } from './common'

export const feedRequests = async (
  context: ReactUseApi.Context,
  ssrHtml: string,
  maxRequests = context.settings.maxRequests
) => {
  const {
    settings,
    collection: { ssrConfigs, cacheKeys },
  } = context
  const { cache, renderSSR, debug } = settings
  if (!ssrConfigs.length) {
    if (debug) {
      console.log('[ReactUseApi][Requests Count] =', cacheKeys.size)
      console.log(
        '[ReactUseApi][Executed Times] =',
        settings.maxRequests - maxRequests
      )
    }
    cacheKeys.clear()
    return ssrHtml // done
  }
  debug && console.log('[ReactUseApi][Collecting Requests]')
  if (maxRequests === 0) {
    console.error(
      '[ReactUseApi][ERROR] - Maximum executing times while fetching axios requests',
      '[congis]:',
      ssrConfigs,
      '[Executed Times]:',
      settings.maxRequests - maxRequests
    )
    cacheKeys.clear()
    return ssrHtml // done
  }

  // The algorithm is collecting the unobtained API request config from the previous renderSSR()
  // , but here only fetches the API data from the first config and again uses renderSSR() to feed the data to its components.
  // This approach may look like inefficient but rather stable, since each config may rely on the data from useApi().
  // However,  it is possible that no one request config that depends on another one, only one renderSSR() is needed
  // , but who can guarantee that every developer is able to consider this dependency?
  // react-apollo uses the similar algorithm
  // https://github.com/apollographql/react-apollo/blob/master/packages/ssr/src/getDataFromTree.ts
  const { config, cacheKey } = ssrConfigs[0] // fetch the first
  const cacheData = cache.get(cacheKey)
  if (!cacheData) {
    try {
      debug && console.log('[ReactUseApi][Fetch]', cacheKey)
      const response = await axiosAll(context, config)
      cache.set(cacheKey, {
        response,
      })
    } catch (error) {
      // is an axios error
      if (error?.response?.data) {
        cache.set(cacheKey, {
          // should not be error (Error) object in SSR, it will lead an error: Converting circular structure to JSON
          error: error.response,
        })
      } else {
        throw error
      }
    }
  }
  // execute renderSSR one after another to get more ssrConfigs
  ssrConfigs.length = 0
  ssrHtml = renderSSR()
  return await feedRequests(context, ssrHtml, --maxRequests)
}

export const injectSSRHtml = async (
  context: ReactUseApi.CustomContext,
  renderSSR?: ReactUseApi.Settings['renderSSR'],
  postProcess?: (ssrHtml: string, apiCacheScript: string) => string
) => {
  context = configure(context, true)
  const { settings } = context
  settings.renderSSR = renderSSR || settings.renderSSR
  const { cache, useCacheData, clientCacheVar, shouldUseApiCache } = settings
  cache.reset()
  // collect API requests first
  let ssrHtml = settings.renderSSR()
  ssrHtml = await exports.feedRequests(context, ssrHtml)
  if (useCacheData) {
    const cacheJson = cache.dump().filter(({ k: cacheKey }) => {
      let config: ReactUseApi.Config
      try {
        config = JSON.parse(cacheKey)
      } catch (e) {
        config = {}
      }
      return shouldUseApiCache(config, cacheKey) !== false
    })
    const apiCacheScript = Object.keys(cacheJson)
      ? `<script>window.${clientCacheVar} = ${JSON.stringify(cacheJson).replace(
          /</g,
          '\\u003c'
        )}</script>`
      : ''
    return isFunction(postProcess)
      ? postProcess(ssrHtml, apiCacheScript)
      : ssrHtml + apiCacheScript
  }
  return ssrHtml
}

export const loadApiCache = (
  context: ReactUseApi.Context = { settings: defaultSettings }
) => {
  const { settings } = context
  const { clientCacheVar } = settings
  const data = window[clientCacheVar]
  if (Array.isArray(data)) {
    settings.cache.load(data)
    delete window[clientCacheVar]
  }
}
