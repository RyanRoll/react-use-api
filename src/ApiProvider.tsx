import React, { useMemo } from 'react'

import { ApiContext } from './context'
import { configure } from './common'

export const ApiProvider: React.FC<ReactUseApi.ApiProviderProps> = ({
  context = {},
  children,
}) => {
  const apiContext = useMemo(() => configure(context), [context])
  return (
    <ApiContext.Provider value={apiContext}>{children}</ApiContext.Provider>
  )
}

ApiProvider.displayName = 'ApiProvider'
