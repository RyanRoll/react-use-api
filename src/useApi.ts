import {
  useEffect,
  useLayoutEffect,
  useReducer,
  useMemo,
  useContext,
  useCallback,
  useRef,
} from 'react'

import { ApiContext } from './context'
import {
  initState,
  ACTIONS,
  axiosAll,
  getResponseData,
  isObject,
  isFunction,
} from './common'

export function useApi<D = ReactUseApi.SingleData>(
  config: ReactUseApi.SingleConfig | string,
  opt?: ReactUseApi.Options | ReactUseApi.Options['handleData']
): [D, ReactUseApi.State, ReactUseApi.RequestFn]
export function useApi<D = ReactUseApi.SingleData[]>(
  config: ReactUseApi.MultiConfigs,
  opt?: ReactUseApi.Options | ReactUseApi.Options['handleData']
): [D, ReactUseApi.State, ReactUseApi.RequestFn]
export function useApi<D = ReactUseApi.Data>(
  config: ReactUseApi.Config | string,
  opt?: ReactUseApi.Options | ReactUseApi.Options['handleData']
): [D, ReactUseApi.State, ReactUseApi.RequestFn] {
  if (typeof config === 'string') {
    config = {
      url: config,
    }
  }
  const context = useContext(ApiContext)
  const {
    settings,
    settings: { cache, debug, clearLastCacheWhenConfigChanges },
    isSSR,
    collection: { ssrConfigs, cacheKeys },
  } = context
  const cacheKey = JSON.stringify(config)
  const ref = useRef(
    useMemo(
      () =>
        ({
          id: Date.now(),
          isRequesting: false,
          isInit: false,
          hasFed: false,
          refreshFlag: 1,
          cacheKey,
          config,
        } as ReactUseApi.RefData),
      []
    )
  )
  const options = useMemo(
    () => handleUseApiOptions(opt, settings, config, cacheKey),
    [opt, settings, cacheKey]
  )
  const { current: refData } = ref
  const isValidConfig = verifyConfig(config)
  const hasChangedConfig = refData.cacheKey !== cacheKey
  options.$hasChangedConfig = hasChangedConfig
  if (hasChangedConfig) {
    if (clearLastCacheWhenConfigChanges) {
      cache.del(refData.cacheKey)
    }
    refData.cacheKey = cacheKey
    refData.config = config
    refData.hasFed = false
  }

  // SSR processing
  const cacheData: ReactUseApi.CacheData = cache.get(cacheKey)
  const { skip, useCache } = options
  let defaultState = { ...initState }

  if (!skip && !refData.isInit) {
    if (cacheData && !refData.hasFed && (isSSR || useCache !== false)) {
      const { response, error } = cacheData
      const action = {
        type: ACTIONS.REQUEST_END,
        response,
        error,
        options,
      } as ReactUseApi.Action
      debug && console.log('[ReactUseApi][Feed]', cacheKey)
      if (!isSSR) {
        action.fromCache = true
        refData.hasFed = true
      }
      defaultState = reducer(defaultState, action)
    } else if (isSSR) {
      if (!cacheKeys.has(cacheKey)) {
        cacheKeys.add(cacheKey)
        debug && console.log('[ReactUseApi][Collect]', cacheKey)
      }
      ssrConfigs.push({
        config,
        cacheKey,
      })
    }
  }

  refData.isInit = true

  const [state, dispatch] = useReducer(reducer, defaultState)
  const { shouldRequest, watch } = options
  const { loading, data } = state

  const request = useCallback(
    async (
      cfg = refData.config as ReactUseApi.Config,
      keepState = false,
      revalidate = true
    ) => {
      if (options.skip) {
        return null
      }
      // foolproof
      if (
        (cfg as React.MouseEvent)?.target &&
        (cfg as React.MouseEvent)?.isDefaultPrevented
      ) {
        cfg = refData.config
      }
      // update state's cachekey for saving the prevState when requesting (refreshing)
      // it's good to set true for pagination
      if (keepState) {
        state.$cacheKey = cacheKey
      }
      refData.isRequesting = true
      return fetchApi(context, cfg, options, dispatch, revalidate)
    },
    [context, cacheKey, options, dispatch, state, refData]
  )

  const shouldFetchApi = useCallback(
    (forRerender = false) => {
      let shouldRequestResult: boolean
      if (isFunction(shouldRequest)) {
        shouldRequestResult = !!shouldRequest() as boolean
      }
      return (
        !skip &&
        !refData.isRequesting &&
        ((forRerender
          ? shouldRequestResult === true
          : // false means skip
            shouldRequestResult !== false) ||
          hasChangedConfig)
      )
    },
    [skip, refData, shouldRequest, hasChangedConfig]
  )

  // for each re-rendering
  if (shouldFetchApi(true)) {
    refData.refreshFlag = Date.now()
  }

  if (!loading && refData.isRequesting) {
    refData.isRequesting = false
  }

  const useIsomorphicLayoutEffect = isSSR ? useEffect : useLayoutEffect
  const effect: typeof useIsomorphicLayoutEffect = useCallback(
    (callback, ...args) => {
      const wrapper = () => {
        if (isValidConfig) {
          return callback()
        }
      }
      return useIsomorphicLayoutEffect(wrapper, ...args)
    },
    [isValidConfig]
  )
  effect(() => {
    // SSR will never invoke request() due to the following cases:
    // 1. There is a cacheData for the cacheKey
    // 2. Feeding the data come from the cache (using defaultState)
    // 3. Calling API forcibly due to useCache=false
    // For non-SSR, cacheData will be undefined if cacheKey has been changed

    if (!isSSR && !refData.hasFed) {
      request(undefined, undefined, false)
    }
  }, [
    cacheKey,
    useCache,
    refData,
    refData.refreshFlag,
    ...(Array.isArray(watch) ? watch : []),
  ])

  if (!isValidConfig) {
    return [undefined, undefined, undefined]
  }
  return [data, state, request]
}

export const reducer = (
  state: ReactUseApi.State,
  action: ReactUseApi.Action
): ReactUseApi.State => {
  const { type, options } = action
  const cacheKey = options.$cacheKey
  const basicState = {
    $cacheKey: cacheKey,
  }
  switch (type) {
    case ACTIONS.REQUEST_START: {
      return {
        // reset the state to initState if the cacheKey is changed on the fly
        ...(cacheKey !== state.$cacheKey ? initState : state),
        loading: true,
        error: undefined,
        fromCache: false,
        ...basicState,
      }
    }
    case ACTIONS.REQUEST_END: {
      const { response, error, fromCache } = action
      const { prevState: pre, ...prevState } = state
      const { data: prevData } = prevState
      const { dependencies, $hasChangedConfig } = options
      const newState = {
        ...prevState,
        prevData,
        prevState,
        loading: false,
        response,
        dependencies,
        error,
        fromCache: !!fromCache,
        ...basicState,
      }
      if ($hasChangedConfig) {
        delete newState.prevState
        delete newState.prevData
      }
      newState.data = error ? undefined : getResponseData(options, newState)
      return newState
    }
    default:
      return state
  }
}

export const fetchApi = async (
  context: ReactUseApi.Context,
  config: ReactUseApi.Config,
  options: ReactUseApi.Options,
  dispatch: React.Dispatch<ReactUseApi.Action>,
  revalidate = false
) => {
  const {
    settings: { cache },
  } = context
  const { $cacheKey: cacheKey, useCache } = options
  const promiseKey = `${cacheKey}::promise`
  let { response, error } = {} as ReactUseApi.CacheData
  try {
    const cacheData = cache.get(cacheKey)
    let promise = cache.get(promiseKey)
    response = cacheData?.response
    error = cacheData?.error
    let fromCache = !revalidate
    if (revalidate) {
      dispatch({ type: ACTIONS.REQUEST_START, options })
      response = await axiosAll(context, config)
    } else if (useCache !== false && promise) {
      dispatch({ type: ACTIONS.REQUEST_START, options })
      response = await promise
    } else if (useCache !== false && (response || error)) {
      // skip ACTIONS.REQUEST_START
    } else if (useCache) {
      dispatch({ type: ACTIONS.REQUEST_START, options })
      promise = axiosAll(context, config)
      // save promise for next coming hooks
      cache.set(promiseKey, promise)
      response = await promise
    } else {
      fromCache = false
      dispatch({ type: ACTIONS.REQUEST_START, options })
      response = await axiosAll(context, config)
    }
    dispatch({ type: ACTIONS.REQUEST_END, response, error, options, fromCache })
  } catch (err) {
    error = err
    dispatch({
      type: ACTIONS.REQUEST_END,
      error,
      options,
    })
  } finally {
    // save the data if there is no cache data come from server
    if (useCache) {
      if (!cache.has(cacheKey)) {
        cache.set(cacheKey, {
          response,
          error,
        })
      }
      if (cache.has(promiseKey)) {
        cache.del(promiseKey)
      }
    }
  }
}

export const handleUseApiOptions = (
  opt:
    | ReactUseApi.Options
    | ReactUseApi.Options['handleData']
    | ReactUseApi.Options['watch'],
  settings: ReactUseApi.Settings,
  config: ReactUseApi.Config | string,
  cacheKey: string
) => {
  const options = isObject(opt)
    ? ({ ...opt } as ReactUseApi.Options)
    : ({
        watch: Array.isArray(opt) ? opt : [],
        handleData: isFunction(opt) ? opt : undefined,
      } as ReactUseApi.Options)

  options.$cacheKey = cacheKey
  const { alwaysUseCache, shouldUseApiCache } = settings
  const globalUseCache =
    shouldUseApiCache(config as ReactUseApi.Config, cacheKey) !== false
  if (alwaysUseCache) {
    options.useCache = true
  } else if (globalUseCache === false) {
    options.useCache = false
  }
  return options
}

export const verifyConfig = (config: ReactUseApi.Config) => {
  return isObject(config)
    ? !!(config as ReactUseApi.SingleConfig).url
    : Array.isArray(config)
    ? config.every((each) => !!each.url)
    : false
}

export default useApi
