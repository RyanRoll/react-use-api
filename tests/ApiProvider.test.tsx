/// <reference path="../src/typings.d.ts" />

import React, { useContext } from 'react'
import LRU from 'lru-cache'
import TestRenderer from 'react-test-renderer'

import { ApiProvider } from '../src'
import { ApiContext } from '../src/context'
import { defaultSettings } from '../src/common'

describe('ApiProvider tests', () => {
  it('should default context work as expected if context does not pass to ApiProvider', () => {
    let context: ReactUseApi.Context = {}
    const App: React.FC = () => {
      context = useContext(ApiContext)
      return <>Hello World</>
    }
    const testRenderer = TestRenderer.create(
      <ApiProvider>
        <App />
      </ApiProvider>
    )
    const { settings, isSSR } = context
    expect(settings).toMatchObject({
      ...defaultSettings,
      cache: new LRU()
    })
    expect(isSSR).toBe(false) // window exists in jest
  })
  it('should context work as expected ', () => {
    const context: ReactUseApi.Context = {}
    const testRenderer = TestRenderer.create(<ApiProvider context={context} />)
    const { root } = testRenderer
    const instance = root.findByType(ApiProvider)
    const { settings, isSSR } = context
    expect(instance.props.context).toBe(context)
    expect(settings).toMatchObject({
      ...defaultSettings,
      cache: new LRU()
    })
    expect(isSSR).toBe(false) // window exists in jest
  })
})
