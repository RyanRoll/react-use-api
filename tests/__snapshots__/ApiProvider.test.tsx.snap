// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiProvider tests should default context work as expected if context does not pass to ApiProvider 1`] = `
<div>
  Hello World
</div>
`;
